### PR TITLE
[AAP-79959] fix: cancel AAP job when scaffolder task is cancelled

### PR DESCRIPTION
**Pull request by abhikdps**
## Description

- Add `cancelJob` method to `AAPClient` that POSTs to the AAP `jobs/{id}/cancel/` endpoint
- Make the polling loop in `aapLaunchJobTemplate` respond to the scaffolder's `AbortSignal` - on cancellation, the AAP job is cancelled before the error propagates
- Add an abort-aware `sleepMs` helper that clears its timer immediately on signal abort (no leaked timeouts)

## Related Issues

Related JIRA: [AAP-79959](https://redhat.atlassian.net/browse/AAP-79959)

## Type of Change

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Documentation update

## Testing

- [x] Signal already aborted before polling enters first iteration
- [x] Signal aborted during polling sleep (mid-flight cancel)
- [x] `cancelJob` API call fails after abort - logs warning, still propagates abort
- [x] `sleepMs` early-reject when signal is pre-aborted
- [x] Polling timeout after MAX_POLLS (720) exceeded
- [x] Non-abort errors from `getJobStatus` re-thrown without calling `cancelJob`
- [x] `AAPClient.cancelJob` - success, network error, 403 forbidden
- [x] `yarn tsc` clean, `yarn lint` clean

## Checklist

- [x] Code follows project style
- [x] Tests pass locally
- [ ] Documentation updated

## Summary by CodeRabbit

* **New Features**
  * Added server-side support to cancel AAP jobs via a new `cancelJob` operation.
  * Improved job template launch behavior with abort-aware completion polling.

* **Bug Fixes**
  * Standardized terminal-state detection for job polling and cancellation to avoid inconsistent status handling.
  * Ensures aborted launches keep the original abort error identity and cancels the running job when appropriate.

* **Tests**
  * Added/expanded coverage for cancel behavior, abort timing edge cases, polling failures/timeouts, and terminal-status validation.

### CODE REVIEW

**Comment by abhikdps** [comment]
@coderabbitai review

File: plugins/scaffolder-backend-module-backstage-rhaap/src/actions/aapLaunchJobTemplate.test.ts [hunk 1]
```diff
@@ -153,6 +153,327 @@ describe('ansible-aap:jobTemplate:launch', () => {
...
+    mockAnsibleService.launchJobTemplateNoWait.mockResolvedValue(
+      launchResponse,
+    );
+    mockAnsibleService.cancelJob.mockResolvedValue(undefined);
+
+    let abortedAccessCount = 0;
+    const mockSignal = {
+      get aborted() {
+        abortedAccessCount++;
+        // First access (line 121 guard) returns false; second (sleepMs line 15) returns true
```
Comment (NilashishC) [inline plugins/scaffolder-backend-module-backstage-rhaap/src/actions/aapLaunchJobTemplate.test.ts]: **Nit**: The line-number references in this comment are stale — "line 121 guard" and "sleepMs line 15" don't correspond to the actual source. The pre-launch guard is at line 147 in the handler and `sleepMs`'s pre-abort check is at line 33.

Consider referencing behavior instead of line numbers to avoid this kind of drift:

```suggestion
        // First access (handler pre-launch guard) returns false; second (sleepMs pre-abort check) returns true
```

File: plugins/backstage-rhaap-common/src/AAPClient/AAPClient.ts [hunk 2]
```diff
@@ -779,6 +780,29 @@ export class AAPClient implements IAAPService {
     }
   }
 
+  public async cancelJob(jobID: number, token: string): Promise<void> {
+    const status = await this.getJobStatus(jobID, token);
```
Comment (NilashishC) [inline plugins/backstage-rhaap-common/src/AAPClient/AAPClient.ts]: **Minor optimization**: `getJobStatus` also fetches all events when the job is in a terminal state (line 770), which adds an unnecessary API round-trip for the cancel path where we only need the status string.

Not blocking — the cancel path is best-effort anyway — but a lighter-weight status-only check could avoid the extra call. Something like a dedicated `getJobStatusOnly` method or just hitting the job detail endpoint directly and checking `status` without fetching events.

File: plugins/scaffolder-backend-module-backstage-rhaap/src/actions/aapLaunchJobTemplate.ts [hunk 3]
```diff
@@ -10,6 +11,96 @@ import {
...
 import { normalizeTemplateLaunchValues } from './schemas/rhaapActionPayloadUtils';
 
+const POLL_INTERVAL_MS = 5000;
+const MAX_POLLS = 720;
+const TERMINAL_STATUSES = new Set([
+  'successful',
+  'failed',
+  'error',
+  'canceled',
+]);
```
Comment (NilashishC) [inline plugins/scaffolder-backend-module-backstage-rhaap/src/actions/aapLaunchJobTemplate.ts]: **Follow-up suggestion**: The terminal status list `['successful', 'failed', 'error', 'canceled']` is duplicated in several places across the codebase — `AAPClient.fetchResult`, `AAPClient.getJobStatus`, `AAPClient.cancelJob`, and now here as a `Set`. Consider extracting this to a shared constant in `backstage-rhaap-common` in a follow-up PR to keep them in sync.

**Review by NilashishC — APPROVED** [review]
Clean, well-tested implementation. The abort-aware polling and `cancelJob` integration work correctly across all the edge cases I checked. A few minor suggestions in the inline comments — nothing blocking. Nice work.

File: plugins/scaffolder-backend-module-backstage-rhaap/src/actions/aapLaunchJobTemplate.test.ts (hunk 1 above)
Comment (abhikdps) [reply plugins/scaffolder-backend-module-backstage-rhaap/src/actions/aapLaunchJobTemplate.test.ts]: updated in commit [9a4f673](https://github.com/ansible/ansible-backstage-plugins/pull/416/commits/9a4f673f5ac36e1abf89ee94e8615949ac4fa17f)

File: plugins/backstage-rhaap-common/src/AAPClient/AAPClient.ts (hunk 2 above)
Comment (abhikdps) [reply plugins/backstage-rhaap-common/src/AAPClient/AAPClient.ts]: Added a private `isJobTerminal()` method that only GETs the job detail endpoint and checks the status. no `fetchEvents` call. `cancelJob` now uses this instead of `getJobStatus`. done as part of commit [9a4f673](https://github.com/ansible/ansible-backstage-plugins/pull/416/commits/9a4f673f5ac36e1abf89ee94e8615949ac4fa17f)

File: plugins/scaffolder-backend-module-backstage-rhaap/src/actions/aapLaunchJobTemplate.ts (hunk 3 above)
Comment (abhikdps) [reply plugins/scaffolder-backend-module-backstage-rhaap/src/actions/aapLaunchJobTemplate.ts]: extracted `TERMINAL_JOB_STATUSES` as a shared `Set` in `backstage-rhaap-common`'s `constants.ts` file and replaced all inline arrays in `AAPClient.ts` and `aapLaunchJobTemplate.ts`. done as part of commit [9a4f673](https://github.com/ansible/ansible-backstage-plugins/pull/416/commits/9a4f673f5ac36e1abf89ee94e8615949ac4fa17f)

**Review by audgirka — APPROVED** [review]
LGTM
